src/check.js

```diff
@@ -40,14 +40,61 @@ function detect(detectors, node, deps) {
     .value();
 }
 
+// Apply imports map from package.json to a discovered dependency.  If the
+// dependency starts with '#' and we have a matching entry in "imports" in
+// the package.json we'll substitute the possible mapped imports in place of
+// that dependency.
+//
+// Conditions can be well-known ones implemented by node, TypeScript, or webpack like
+// "import", "browser", "types", or "require".  They can also be custom ones as configurable
+// in webpack configuration or using the enhanced-resolve package.
+//
+// See also:
+//   - https://nodejs.org/api/packages.html#subpath-imports
+//   - https://www.typescriptlang.org/docs/handbook/esm-node.html
+//   - https://webpack.js.org/configuration/resolve/#resolveconditionnames
+function applyImportsMap(importsMap, dep) {
+  const resolvedDeps = [];
+
+  function accumulateDeps(v) {
+    if (v) {
+      if (typeof v === 'string') {
+        resolvedDeps.push(requirePackageName(v));
+      } else if (typeof v === 'object') {
+        Object.values(v).forEach(accumulateDeps);
+      }
+    }
+  }
+
+  if (dep.startsWith('#')) {
+    Object.entries(importsMap).forEach((m) => {
+      if (`${dep}/`.startsWith(`${m[0]}/`)) {
```

Comment:
Can't this be simplified? 

```suggestion
      if (dep.startsWith(m[0])) {
```

Reply:
I'm trying to avoid a case where you have, say:

```
"imports": {
  "#foobar": "some-other-package"
  "#foo": "some-package",
}
```

If you import `#foo` or `#foo/whatever` it should only match a key of `#foo` with no extra characters after *except* a `/`.

Reply:
Maybe I wrote it in a confusing way, though ... maybe I need to split on slash and compare explicitly 🤔

Reply:
I rewrote it to be a lot more clear what is going on.

Reply:
Ah, right, but then even adding a `/` won't do us no good. You can put 

```
  "imports": {
    "#toto": "./file1.js",
    "#toto/hm": "./file2.js"
  }
```
in your package.json file and in this case depcheck could be wrong no?

Do we even need a `startsWith`? Shouldn't the imports in the code be exactly the same as defined in package.json?

Reply:
Right so I just checked something, I hope I'm checking the right thing, it's getting late here :D 

With `"#ho": "./ha"` in my imports in package.json, and `ha` being a directory I can't `require('#ho/pf.js'), I get an error:

```
TypeError [ERR_PACKAGE_IMPORT_NOT_DEFINED]: Package import specifier "#ho/pf.js" is not defined in package /private/tmp/ntest/package.json imported from /private/tmp/ntest/index.js
```

Reply:
Also, you could have 

```
"#ho": "./foo.js"
"#ho/pf": "./ha/pf.js"
```
in your imports.

So I'm beginning to think this should really be a `===` check.

Reply:
Ah, true, actually to import a subpath generically it would be more like:

```
"imports": {
"#ho/*": "ha/*"
}
```

So to handle this exactly to spec I'd have to do the wildcard processing.  Hmm.

Reply:
Ha, indeed wildcard works! Is there a real spec somewhere? I didn't see any of that in the PR description. If it's a regex then why don't we check it as a regex?

If we `dep.match(m)` then there is no more `#foo/thing` and `#foobar` shenanigans.

Reply:
Right, so it seems only `*` is permitted? https://nodejs.org/api/packages.html#subpath-patterns

Reply:
For future people reading this, here is what node is doing

https://github.com/nodejs/node/blob/4e0184294362b477e52c996d304e6fe0d6824964/lib/internal/modules/esm/resolve.js#L612-L670

Reply:
I updated it to handle wildcards the same way as documented in the node docs (I hope!).

Reply:
Or at least, hopefully it's close enough to avoid most false reports of unused deps that are actually used indirectly via this feature.

Reply:
Yeah, I'm OK with the latest code you pushed, it's might not be 100% in line with node but at least it fixes a whole class of errors we had. If people stumble on a difference between the implementations they will (hopefully) open issues :)

Thanks for working on this 👍